### PR TITLE
Attribution Chat UI

### LINKED ISSUE #59335: IntelliJ > Guardrails Chat UI

### PR DESCRIPTION
**Pull request by cbart**
Closes sourcegraph/sourcegraph#59335

This pull-request implements attribution search in Chat:

- Extend the agent API (JetBrains side) to call `attribution/search` (see https://github.com/sourcegraph/cody/pull/2798)
- Display a disabled button on hover of chat code editor that presents that status of attribution search **text** (*tooltip*):
  - **Attribution search** () - when still awaiting on whole snippet to be typed in.
  - **Attribution search** (*Guard Rails: Running Code Attribution Check...*) once snippet is done typing and search is running.
  - **Guard Rails API Error** (*Guard Rails API Error: ...*) if attribution search is unavailable
  - **Guard Rails Check Passed** (*Snippet not found on Sourcegraph.com.*) if attribution search finished with no results.
  - **Guard Rails Check Failed** (*Guard Rails Check Failed. Code found in %d repositories ...*) if attribution search finished and attribution was found.

Synopsis of main participating classes (in order of appearance):
* `AttributionSearchCommand` starts attribution search and then notifies UI of the result.
* `AttributionButtonController` drives UI changes of the attribution button.
* `ConditionalVisibilityButton` allows for showing the attribution button on hover (`visible`) only if attribution is enabled (`visibilityAllowed`)

Besides the UI changes to arrange everything neatly are a little scattered around.

## User interface

Visually the indicator is a disabled button to match the current copy/paste at cursor buttons. @danielmarquespt is working on design to match the VS Code UI layout.

**Check pending**

pending

**Check passed**

passed

**Check failed**

failed

## Test plan

- [x] Manual testing
- [ ] Unit tests
- [ ] (stretch) Acceptance Heavy UI test

I gave up on trying to write a component or acceptance test here, since we really need to do the heavy lifting of making this extension testable. This work is beyond the scope of this issue and the timelines we have at this point, but I will come back to that (https://github.com/sourcegraph/sourcegraph/issues/60136)

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/ui/AttributionButtonController.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          project.getService(CurrentConfigFeatures::class.java)
+      // Only display the button if attribution is enabled.
+      button.visibilityAllowed = currentConfigFeatures.get().attribution
+      return AttributionButtonController(button)
+    }
+  }
+
+  @RequiresEdt
+  override fun onAttributionSearchStart() {
+    button.toolTipText = "Guard Rails: Running Code Attribution Check..."
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/ui/AttributionButtonController.kt]: nit: Can you please bundle these tool tips texts and texts? 🙏

File: src/main/kotlin/com/sourcegraph/cody/ui/AttributionButtonController.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import com.sourcegraph.cody.agent.protocol.AttributionSearchResponse
+import com.sourcegraph.cody.attribution.AttributionListener
+
+class AttributionButtonController(val button: ConditionalVisibilityButton) : AttributionListener {
+
+  private val extraUpdates: MutableList<Runnable> = ArrayList()
+
+  companion object {
+    fun setup(project: Project): AttributionButtonController {
+      val button = ConditionalVisibilityButton("Attribution search")
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/ui/AttributionButtonController.kt]: What about the chat persistence? when I reopen the chat after IDE restart I see this text only 😕

File: src/main/kotlin/com/sourcegraph/cody/ui/AttributionButtonController.kt (hunk 1 above)
Comment (cbart) [reply src/main/kotlin/com/sourcegraph/cody/ui/AttributionButtonController.kt]: Done!

**Review by mkondratek — APPROVED** [review]

File: src/main/kotlin/com/sourcegraph/cody/ui/AttributionButtonController.kt (hunk 2 above)
Comment (cbart) [reply src/main/kotlin/com/sourcegraph/cody/ui/AttributionButtonController.kt]: Filed https://github.com/sourcegraph/sourcegraph/issues/60229

**Comment by cbart** [comment]
All comments addressed, and tested manually again. I've updated the failed tooltip to use ul/li for displaying repos, since that seemed to render better:

Screenshot 2024-02-06 at 9 22 25 PM

**Comment by mkondratek** [comment]
It looks much better now 🚀

**Comment by cbart** [comment]
https://www.youtube.com/watch?v=HKK4KmDlj8U

**Pull request merged**